utils/bfd.ml

```diff
@@ -0,0 +1,540 @@
+    seek Int64.(add e_lfanew (of_int (header_size + size_of_optional_header)));
+    let mk buf =
+      let name                = name_at ~max:8 buf 0 in
+      let virtual_size        = get_int   buf 8 in
+      let virtual_address     = get_int32 buf 12 in
+      let size_of_raw_data    = get_int   buf 16 in
+      let pointer_to_raw_data = get_int   buf 20 in
+      {name; virtual_size; virtual_address; size_of_raw_data; pointer_to_raw_data}
+    in
+    Array.init number_of_sections (fun _ -> mk (read_bytes section_header_size))
+
+  type symbol =
+    {
+      name: string;
+      address: int64;
+    }
+
+  let read_section_body {size_of_raw_data; pointer_to_raw_data; _} =
+    seek (Int64.of_int pointer_to_raw_data);
+    read_bytes size_of_raw_data
+
+  let find_section sectname =
+    array_find (function ({name; _} : section) -> name = sectname) sections
+
+  let symbols =
+    match find_section ".exptbl" with
+    | None -> [| |]
+    | Some ({virtual_address; _} as exptbl) ->
+        let data       = read_section_body exptbl in
+        let numexports = Int64.to_int (get_word data 0) in
+        Array.init numexports (fun i ->
+            let address = get_word data (word_size + word_size * 2 * i) in
+            let nameoff = get_word data (word_size + word_size * 2 * i + word_size) in
+            let name    =
+              let off =
+                Int64.(sub nameoff (add (unsigned_of_int32 virtual_address) image_base))
+              in
+              name_at data (Int64.to_int off)
+            in
+            {name; address}
+          )
+
+  let symbol_offset symname =
+    match array_find (function {name; _} -> name = symname) symbols with
+    | None -> None
+    | Some {address; _} ->
+        Some Int64.(sub address (add (unsigned_of_int32 base_of_code) image_base))
+
+  let defines_symbol symname =
+    Array.exists (fun {name; _} -> name = symname) symbols
+end
+
+module type T = functor (_ : READ) -> S
+
+let guess ic =
+  seek_in ic 0;
+  let magic = really_input_string ic 4 in
+  match magic.[0], magic.[1], magic.[2], magic.[3] with
+  | '\x7F', 'E', 'L', 'F' ->
+      (module ELF : T)
+  | '\xFE', '\xED', '\xFA', '\xCE'
+  | '\xCE', '\xFA', '\xED', '\xFE'
+  | '\xFE', '\xED', '\xFA', '\xCF'
+  | '\xCF', '\xFA', '\xED', '\xFE' ->
+      (module Mach_O : T)
+  | 'M', 'Z', _, _ ->
+      (module FlexDLL : T)
+  | _ ->
+      raise (Error (Unrecognized magic))
+
+type t =
+  (module S)
+
+let with_open_in fn f =
+  let ic = open_in_bin fn in
+  Fun.protect ~finally:(fun () -> close_in_noerr ic) (fun () -> f ic)
+
+let read ic =
+  let module R = struct
+    let read_bytes len =
+      let buf = Bytes.create len in
+      really_input ic buf 0 len;
+      buf
+    let seek pos =
+      LargeFile.seek_in ic pos
+    let advance n =
+      LargeFile.seek_in ic (Int64.add (LargeFile.pos_in ic) (Int64.of_int n))
+  end in
+  let (module O : T) = guess ic in
+  (module O (R) : S)
+
+let read filename =
```

Comment:
Personally I would catch the `Error` exception and transform that to a `result` type.

Reply:
OK. Wish granted.